vault/request_handling.go

```diff
@@ -0,0 +1,425 @@
+		if !cubbyholeMounted {
+			if c.router.MatchingMount("cubbyhole/") != "" {
+				cubbyholeMounted = true
+			}
+		}
+		cubbyholeMountedMutex.Unlock()
+	}
+
+	// We are wrapping if there is anything to wrap (not a nil response) and a
+	// TTL was specified for the token, plus if cubbyhole is mounted (which
+	// will be the case normally)
+	wrapping := cubbyholeMounted && resp != nil && resp.WrapInfo != nil && resp.WrapInfo.TTL != 0
+
+	// If we are wrapping, the first part happens before auditing so that
+	// resp.WrapInfo.Token can contain the HMAC'd wrapping token ID in the
+	// audit logs, so that it can be determined from the audit logs whether the
+	// token was ever actually used.
+	if wrapping {
+		// Create the wrapping token
+		te := TokenEntry{
+			Path:         req.Path,
+			Policies:     []string{"cubbyhole-response-wrapping"},
+			CreationTime: time.Now().Unix(),
+			TTL:          resp.WrapInfo.TTL,
+			NumUses:      1,
+		}
+
+		if err := c.tokenStore.create(&te); err != nil {
+			c.logger.Printf("[ERR] core: failed to create wrapping token: %v", err)
+			return nil, ErrInternalError
+		}
+
+		resp.WrapInfo.Token = te.ID
+
+		httpResponse := logical.SanitizeResponse(resp)
+
+		cubbyReq := &logical.Request{
+			Operation:   logical.CreateOperation,
+			Path:        "cubbyhole/response",
+			ClientToken: te.ID,
+			Data: map[string]interface{}{
+				"response": httpResponse,
+			},
+		}
+
+		cubbyResp, err := c.router.Route(cubbyReq)
+		if err != nil {
+			// Revoke since it's not yet being tracked for expiration
+			c.tokenStore.Revoke(te.ID)
+			c.logger.Printf("[ERR] core: failed to store wrapped response information: %v", err)
+			return nil, ErrInternalError
+		}
+		if cubbyResp != nil && cubbyResp.IsError() {
+			c.tokenStore.Revoke(te.ID)
+			c.logger.Printf("[ERR] core: failed to store wrapped response information: %v", cubbyResp.Data["error"])
+			return cubbyResp, nil
+		}
+
+		auth := &logical.Auth{
+			ClientToken: te.ID,
+			Policies:    []string{"cubbyhole-response-wrapping"},
+			LeaseOptions: logical.LeaseOptions{
+				TTL:       te.TTL,
+				Renewable: false,
+			},
+		}
+
+		// Register the wrapped token with the expiration manager
+		if err := c.expiration.RegisterAuth(te.Path, auth); err != nil {
+			// Revoke since it's not yet being tracked for expiration
+			c.tokenStore.Revoke(te.ID)
+			c.logger.Printf("[ERR] core: failed to register cubbyhole wrapping token lease "+
+				"(request path: %s): %v", req.Path, err)
+			return nil, ErrInternalError
+		}
+	}
+
+	// Create an audit trail of the response
+	if err := c.auditBroker.LogResponse(auth, req, resp, err); err != nil {
+		c.logger.Printf("[ERR] core: failed to audit response (request path: %s): %v",
+			req.Path, err)
+		return nil, ErrInternalError
+	}
+
+	// If we are wrapping, now is when we create a new response object with the
+	// wrapped information, since the original response has been audit logged
+	if wrapping {
```

Comment:
So this is a bit odd, since the "response" we logged is not actually the response. We may want to instead log two different responses, so that neither is lost.